### ISSUE TITLE
Panics in destructors can cause the return value to be leaked

### ISSUE THREAD

**Issue by arielb1** [rust-lang/rust#47949]
## STR

```Rust
struct NoisyDrop;
impl Drop for NoisyDrop {
    fn drop(&mut self) {
        println!("dropping a NoisyDrop");
    }
}

impl NoisyDrop {
    fn new() -> Self {
        println!("creating a NoisyDrop");
        NoisyDrop
    }
}

struct PanickyDrop;
impl Drop for PanickyDrop {
    fn drop(&mut self) {
        panic!()
    }
}

fn foo() -> NoisyDrop {
    let p = PanickyDrop;
    NoisyDrop::new()
}

fn main() {
    foo();
}
```

## Expected Result

"creating a NoisyDrop" and "dropping a NoisyDrop" should appear the same number of times

## Actual Result

```
creating a NoisyDrop
thread 'main' panicked at 'explicit panic', src/main.rs:18:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

The destructor is ignored.

**Comment by jonas-schievink** [comment]
Heh, this even happens on Rust 1.0

**Comment by arielb1** [comment]
@jonas-schievink 

Rust 1.0 is *far* worse than this - it leaks *everything* when a destructor panics.

**Comment by solb** [comment]
@arielb, could you explain why this is tagged as a soundness bug? The language makes no guarantee that destructors will never be invoked.

**Comment by gnzlbg** [comment]
ping @rust-lang/compiler this should be triagged, even if it is not a soundness bug, unwinding working properly is something that a lot of code relies on.

**Comment by nikomatsakis** [comment]
So, to be clear, it's not that we leak **all** destructors when panic in drop occurs. For example, [this function drops both `p` and `n` as expected](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=495eafee429b7008b235775c9e301d14):

```rust
fn foo() {
    let n = NoisyDrop::new();
    let p = PanickyDrop;
}
```

The specific scenario here is that:

- we are in the middle of returning the `NoisyDrop` to our caller (so our caller "owns it")
- but we panic before our return completes (so they don't realize they own it)

I don't think it's quite right to say that "we make no guarantee that destructors will ever be invoked". We don't guarantee that all destructors are always invoked, but we do guarantee some things: for example, that if you have a local variable whose scope includes a call, it will be dropped when the call panics (here, `n` would be dropped):

```rust
let n = ...;
foo();
```

I actually think that `I-unsound` is appropriate here. The language does not guarantee that destructors will execute, but we *ought to* (I think) guarantee that destructors for your local variables will execute when we unwind. That is to say, we don't guarantee that destructors run in general because values can be leaked -- but if a value is not leaked, then it should be properly dropped. For example, Rayon relies on this. Presumably a "panic in drop" from some user's function

**Comment by nikomatsakis** [comment]
Adding T-lang -- I think that documenting the expected order would be a good first step.

**Comment by nikomatsakis** [comment]
We discussed in the @rust-lang/lang meeting. Everyone agreed this behavior is wrong; the returned value ought to be dropped, presumably after all other variables in the frame have been dropped. Basically it should be equivalent to having stored the value into an outer let.

However, as a good first step, it seems like we should begin by verifying the behavior with a more thorough set of test cases. For example, I found ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=71a7606326adf341f7544e0732aa01b7)) that the problem applies not only to returns but to any store:

```rust
fn foo() -> NoisyDrop {
    let n = NoisyDrop::new(1);
    let x = {
        let p = PanickyDrop;
        NoisyDrop::new(2) // <-- dtor for this never runs 
    };
    panic!()
}
```

Ideally what we would do is this:

- Create various tests covering these corner cases (also perhaps things like panic when running the dtor of a struct field) so we know the **current behavior**
- Decide on the **desired behavior** and, depending on what we think, maybe writing up an RFC
- Implementing the fixes in drop elaboration

We thought perhaps @matthewjasper might be interested in tackling this, in particular =)

**Comment by nikomatsakis** [comment]
I'm leaving nominated so we follow up next meeting and discuss priority -- how serious do we think it is to fix this. Obviously, this is a long-standing behavior.

**Comment by matthewjasper** [comment]
Oh, there's an issue for this.

> * Create various tests covering these corner cases (also perhaps things like panic when running the dtor of a struct field) so we know the **current behavior**

This should be done for evaluation order as well.

> * Implementing the fixes in drop elaboration

The bug is in MIR construction. I don't think it's that hard to fix (except possibly for the return place).

> We thought perhaps @matthewjasper might be interested in tackling this, in particular =)

=)

**Comment by nikomatsakis** [comment]
@matthewjasper 

> The bug is in MIR construction.

Oh?

**Comment by matthewjasper** [comment]
The latest perf results in https://github.com/rust-lang/rust/pull/66703#issuecomment-593002683 are a bit higher than I would like, but I can't see any way to reduce it.

**Comment by RalfJung** [comment]
So https://github.com/rust-lang/rust/pull/61430 was marked as fixing this, but actually did not, or why did the issue get reopened?

**Comment by matthewjasper** [comment]
#61430 was reverted for being far too large a perf regression.

**Comment by RalfJung** [comment]
Oh, looks like it [was backed out due to a perf regression](https://github.com/rust-lang/rust/pull/61430#issuecomment-540175085).

I agree we should avoid perf regressions, but ultimately I think correct compilation is more important than performance... in particular I do think this qualifies as a soundness bug (like all incorrect complications do); there might be unsafe code relying on rustc doing the right thing here.

**Comment by RalfJung** [comment]
Here's an [example for why this is a soundness bug](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=066727d2e50405c4703854a1025ce311). `weird_drop::drop` is silly, but it should be correct, and this bug makes it cause UB (as you can see when running this code in Miri).

I'll re-add the I-unsound label.

**Comment by nikomatsakis** [comment]
Hmm, yeah, I had forgotten about this.

**Comment by Diggsey** [comment]
@RalfJung you predicted correctly, https://github.com/taiki-e/pin-project/pull/194 ran into this issue, and if @taiki-e hadn't spotted it, it would have caused unsoundness in the generated code.

**Comment by RalfJung** [comment]
@Diggsey @taiki-e  that's a good catch!

So, can we revive #66703? I don't know what the policy is on how much perf impact we are willing to accept for soundness fixes -- this is probably a case-by-case thing.

**Comment by jonas-schievink** [comment]
@RalfJung It looks like Matthew is now going for a different approach in #71840 that should be more efficient

**Comment by matthewjasper** [comment]
It's the same approach as in #66703, which has perf runs.

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/71840 had to be reverted unfortunately to fix https://github.com/rust-lang/rust/issues/72470, it seems.

**Comment by matthewjasper** [comment]
Indeed, I'll create an issue to reapply it.

**Comment by Aaron1011** [comment]
https://github.com/rust-lang/rust/pull/71840 has been re-landed in https://github.com/rust-lang/rust/pull/77466. What's the next step for this issue?

**Comment by matthewjasper** [comment]
The next step is for me to resurrect #72152 and see how a perf run looks.

**Comment by pnkfelix** [comment]
Reopening due to PR #81257

**Comment by lcnr** [comment]
This has a conceptionally clear fix by changing MIR building which does not cause any issues, expect a single borrowck regression in #80949. It has also been encountered in the wild at least once in https://github.com/rust-lang/rust/issues/47949#issuecomment-623120647. Going to bump this to `P-high`.

I personally would be in favor of merging https://github.com/rust-lang/rust/pull/78373 again, even if it still breaks `euca` if it is difficult to avoid that issue.

**Comment by nikomatsakis** [comment]
I agree we should fix this. I don't recall the solution in detail nor the issue though. @lcnr it seems to me that FCP is appropriate -- this is likely a @rust-lang/types FCP with @rust-lang/lang cc'd?

**Comment by lcnr** [comment]
An FCP to merge #78373 again, accepting the regressions caused by it, assuming no additional major crates started to depend on the current behavior?

hmm, not totally sure which team is the most responsible here. I feel like it may be types > @rust-lang/compiler > lang as it feels more like an implementation detail/bug than an actual lang design question🤔

Yes, that seems good to me. Will put this on my list of things I'd like to do, though I would love for someone else to summarize this for an FCP proposal, as I may not get to it in the near future.

**Comment by estebank** [comment]
@nikomatsakis this is something I explored with the panic reachability logic in redpen. I can detect that a panic *might* happen in a `Drop` impl, but for a `clippy` or `rustc` lint we'd need something a bit stronger where we *know for a fact* that the panic can hit (it's not just dead unreachable code).

**Comment by nikomatsakis** [comment]
@lcnr I agree with that ordering.

**Comment by theemathas** [comment]
Possibly related to #156713.

**Comment by theemathas** [comment]
A leak can also happen with a closure that panics upon drop, even without any relevant blocks.

```rust
struct LoudDrop;
impl Drop for LoudDrop {
    fn drop(&mut self) {
        println!("dropped");
    }
}
fn new() -> LoudDrop {
    println!("created");
    LoudDrop
}

struct PanicOnDrop;
impl Drop for PanicOnDrop {
    fn drop(&mut self) {
        panic!()
    }
}

fn foo(f: impl FnOnce(LoudDrop) -> LoudDrop) {
    f(new());
}

fn main() {
    let x = PanicOnDrop;
    foo(move |y| (&x, y).1);
}
```

The above code prints "created" but not "drop".
